core/runtime.rs

```diff
@@ -1792,17 +1838,30 @@ impl JsRuntime {
   fn resolve_async_ops(&mut self, cx: &mut Context) -> Result<(), Error> {
     let state_rc = Self::state(self.v8_isolate());
 
-    let js_recv_cb_handle = state_rc.borrow().js_recv_cb.clone().unwrap();
-    let scope = &mut self.handle_scope();
+    // We keep a list of promise IDs and OpResults per realm. Since v8::Context
+    // isn't hashable, `results_per_realm` is a vector of (context, list) tuples
+    type ResultList = Vec<(i32, OpResult)>;
+    let mut results_per_realm: Vec<(v8::Global<v8::Context>, ResultList)> = {
+      let known_realms = &mut state_rc.borrow_mut().known_realms;
+      let mut results = Vec::with_capacity(known_realms.len());
+
+      // Avoid calling the method multiple times
+      let isolate = self.v8_isolate();
+
+      // Remove GC'd realms from `known_realms` at the same time as we populate
+      // `results` with those that are still alive.
+      known_realms.retain(|weak| {
+        if !weak.is_empty() {
+          let context = weak.to_global(isolate).unwrap();
```

Comment:
This seems costly - isn't upgrading weak handle to a global similar to cloning a global?

Reply:
`v8::Weak<T>` is a different Rust wrapper for the C++ `v8::Global<T>`, so yeah, but even if we make `results_per_realm` be `Vec<(v8::Weak<v8::Context>, ResultList)>`, the weaks would have to be cloned in order to add them to the vec.

In the end some cloning or upgrading is necessary, although with something like (ab)using V8's internal handle representation to essentially build a hashmap with `v8::Context` keys, we'd only need to clone globals for the realms that have async ops to resolve. Is this preferable?